providers/fips/selftest.c

```diff
@@ -88,10 +154,30 @@ int SELF_TEST_post(SELF_TEST_POST_PARAMS *st)
     BIO *bio_module = NULL, *bio_indicator = NULL;
     unsigned char *module_checksum = NULL;
     unsigned char *indicator_checksum = NULL;
+    int loclstate;
+
+    if (!RUN_ONCE(&fips_self_test_init, do_fips_self_test_init))
+        return 0;
+
+    CRYPTO_THREAD_read_lock(self_test_lock);
+    loclstate = FIPS_state;
+    CRYPTO_THREAD_unlock(self_test_lock);
 
+    if (loclstate == FIPS_STATE_RUNNING)
+        return 1;
+    if (loclstate != FIPS_STATE_INIT)
+        return 0;
+
+    CRYPTO_THREAD_write_lock(self_test_lock);
+    if (FIPS_state == FIPS_STATE_RUNNING) {
+        CRYPTO_THREAD_unlock(self_test_lock);
+        return 1;
+    }
+    if (FIPS_state != FIPS_STATE_INIT) {
+        CRYPTO_THREAD_unlock(self_test_lock);
+        return 0;
+    }
```

Comment:
Lines 171 to here, I don't quite understand.  Isn't it sufficient to just `CRYPTO_THREAD_write_lock(self_test_lock);`?

Reply:
Something may have changed in between checking the state with the read lock, and re-obtaining it with a write lock. Therefore we have to check the state again.

Reply:
Would wrapping these tests into their own function make sense?
Rather than duplicating the block of code.